ratelimitstore.go

```diff
@@ -0,0 +1,94 @@
+package desync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
+)
+
+type ThrottleOptions struct {
+	eventRate                float64
+	burstRate int
+
+
+}
+
+
+type RateLimitedStore struct {
+
+	wrappedStore WriteStore
+
+	limiter  *rate.Limiter
+
+	options ThrottleOptions
+
+}
+
+var RateLimitedExceeded = errors.New("Rate Limit Exceeded")
+
+
+func NewRateLimitedStore(s WriteStore, options ThrottleOptions) *RateLimitedStore {
+	
+	limiter := rate.NewLimiter(rate.Limit(options.eventRate), options.burstRate)
+	return &RateLimitedStore{wrappedStore: s,limiter: limiter, options: options }
+}
+
+func (s RateLimitedStore) GetChunk(id ChunkID) (*Chunk, error) {
+
+	chunk,err := s.wrappedStore.GetChunk(id)
+	if err != nil{
+		return chunk, err
+	}
+	ctx := context.Background()
+
+	err  = s.limiter.WaitN(ctx,1)
+	return chunk, err
+}
+
+func (s RateLimitedStore) HasChunk(id ChunkID) (bool, error) {
+
+	
+
+	has,err := s.wrappedStore.HasChunk(id)
+	if err != nil{
+		return has, err
+	}
+	ctx  :=context.Background()
+	err  = s.limiter.WaitN(ctx,1)
+	return has, err
+	
+}
+
+
+func (s RateLimitedStore) StoreChunk(chunk *Chunk) error {
+	
+	// This isn't ideal because what I'm really interested is in size over the wire.
```

Comment:
I think you can actually achieve this by using a limit of "bytes per second" (must be equal or larger than max-chunk-size). The here in `StoreChunk()` you can
```go
	b, err := chunk.Data()
	if err != nil {
		return err
	}
	err = s.limiter.WaitN(ctx, len(b))
	if err != nil {
		return RateLimitedExceeded
	}
        return s.wrappedStore.StoreChunk(chunk)
```

It's a little more interesting in `GetChunk()` since you don't know how large it is. One way would be to `WaitN(ctx, <avg-chunk-size>)` before calling the wrapped store, or `WaitN(ctx, <size of data>)` after actually pulling the chunk from the wrapped store. The latter can cause spikes early on when concurrency is used, but should average to the rate you set over time.

As for `HasChunk()`, not sure if you really need to limit it, it's very small. You can just use `1` like you have now if you want to limit it too, could do that before calling the wrappedStore.